src/solution.cpp

```diff
@@ -394,47 +394,12 @@ void Solution::RunImpl(Handle& handle,
         }
     }();
 
-    if(invoker)
-    {
-        (*invoker)(handle, invoke_ctx);
-        return;
-    }
-
-    solver::mha::MhaForward mhaForward;
-    solver::mha::MhaBackward mhaBackward;
-
-    if(!kernels.empty())
-    {
-        const auto ctx          = ExecutionContext{&handle};
-        const auto mha_solution = GetSolver() == mhaForward.SolverDbId()
-                                      ? mhaForward.GetSolution(ctx, problem_description)
-                                      : mhaBackward.GetSolution(ctx, problem_description);
-        auto kernel_handles     = std::vector<Kernel>{std::begin(kernels), std::end(kernels)};
-
-        invoker = (*mha_solution.invoker_factory)(kernel_handles);
-        (*invoker)(handle, invoke_ctx);
-        return;
-    }
-
-    const auto net_cfg = problem_description.MakeNetworkConfig();
-    invoker            = handle.GetInvoker(net_cfg, GetSolver());
-
-    if(invoker)
-    {
-        (*invoker)(handle, invoke_ctx);
-        return;
-    }
-
-    auto ctx = ExecutionContext{&handle};
-
-    const auto mha_solution = GetSolver() == mhaForward.SolverDbId()
-                                  ? mhaForward.GetSolution(ctx, problem_description)
-                                  : mhaBackward.GetSolution(ctx, problem_description);
+    const auto algo    = AlgorithmName{"MHA"};
```

Comment:
In Problem.cpp file we have this name as "Mha". Here it's "MHA" with all capitals. Should we stick with the same name,  "Mha" ?

Reply:
Also, are you sure about the correctness of such a code replacement in general?
In RunImpl you try to find a solution again instead of using the solution which we already found on FindSolution phase. It may happen that ExecutePrimitive will execute not the solution we want to execute, because it will execute the first isApplicable solution.

Reply:
Oh good point, ill fix that.

Reply:
hmm, another my comment gone, lol

Reply:
Ill take a look.

Thanks for the heads up!

Reply:
Yea you are correct, this needs to use the solverId, and can't make a new container.

Fixing it!

Reply:
Alright should be updated to be correct now @Vsevolod1983. 
If you get a chance could you take another look at the updated version?

Thanks again for the heads up!